### PR TITLE
Change the default order of data storage objects

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michal Zientkiewicz <michalz@nvidia.com>

## Description:
Change the default order for data storage objects to `host` - regardless if they contain any data or not.

### Tests:

- [x] Existing tests apply
The performance in RN50 and retina was validated. 
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/pipeline/data/tensor_list.cc [hunk 1]
```diff
@@ -463,6 +463,8 @@ const TensorListShape<> &TensorList<Backend>::shape() const & {
 
 template <typename Backend>
 void TensorList<Backend>::set_order(AccessOrder order, bool synchronize) {
+  if (!order)
```
Comment (stiepan) [inline dali/pipeline/data/tensor_list.cc:466]: So you cannot nullify the order of the tensor list?

Comment (stiepan) [reply dali/pipeline/data/tensor_list.cc:466]: And.. what do we need `AccessOrder::null_stream()` for?

**Comment by klecki** [comment]
!build

File: dali/pipeline/data/tensor_list.cc (hunk 1 above)
Comment (klecki) [reply dali/pipeline/data/tensor_list.cc]: I added a restriction that disallows setting the empty order to TensorList and Buffer, and allowed it to mean no-op in the APIs where it was used that way (so it can be an optional argument). For the null_stream, I think it is invalid, and if used I guess we will get an error?

**Comment by klecki** [comment]
!build

File: dali/test/python/test_dali_cpu_only.py [hunk 2]
```diff
@@ -60,6 +60,20 @@ def get_data():
     return out
 
 
+# The same code is used as CPU-only pipeline to test if TF plugin loads successfully
+# during its installation.
+def test_tensorflow_build_check():
```
Comment (stiepan) [inline dali/test/python/test_dali_cpu_only.py:65]: Is this related to the PR?

**Review by stiepan — APPROVED** [review]

File: dali/pipeline/data/buffer.h [hunk 3]
```diff
@@ -492,7 +491,9 @@ class DLL_PUBLIC Buffer {
    *       set_order(order) separately for less synchronization.
    */
   void reset(AccessOrder order = {}) {
-    set_order(order);
+    if (order) {
+      set_order(order);
+    }
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/pipeline/data/buffer.h:496]: In most cases reset means that we free the allocation and you can change the properties, but we kept pinnedness, order and device_id intact.

File: dali/test/python/test_dali_cpu_only.py (hunk 2 above)
Comment (klecki) [reply dali/test/python/test_dali_cpu_only.py:65]: It is a bit related, because I had to run this code to see where it is failing :P

**Comment by klecki** [comment]
!build
